In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
#read data into a pandas dataframe = data
data = pd.read_csv('fuel_ferc1.csv')
data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

## **Question 3: The year with the highest average fuel cost per unit delivered**

In [5]:
#split data into groups ('report_year', 'fuel_cost_per_unit_delivered') using groupby function.
#Aggregate using mean to find average and store in a variable = new_data.
new_data = data[['report_year', 'fuel_cost_per_unit_delivered']].groupby('report_year').mean()

#sort values in descending order so highest value comes first in the dataframe
new_data.sort_values('fuel_cost_per_unit_delivered', ascending=False)

fuel_cost_per_unit_delivered
report_year                              
1997                         11140.197239
1996                          9196.705948
2000                           985.362877
2009                           652.694163
2018                           499.269966
2015                           326.535511
1998                           287.154420
2014                           192.737183
2013                           172.307591
2004                           139.524275
2016                           103.901761
2010                            91.862105
1994                            63.636060
2012                            60.994502
2001                            60.050396
2011                            59.774667
2008                            58.588197
2003                            55.663493
2002                            47.594361
2017                            46.196861
2007                            43.325023
2005                            41.438184
2006                            38.657484
1995                            32.735269
1999                            25.551627

In [14]:
#Get first value
answer = new_data.sort_values('fuel_cost_per_unit_delivered', ascending=False).iloc[0]
answer

fuel_cost_per_unit_delivered    11140.197239
Name: 1997, dtype: float64

## **Question 9: The fuel type code with the lowest fuel cost per unit burned.**

In [7]:
#split data into groups ('fuel_type_code_pudl','fuel_cost_per_unit_burned') using groupby function.
#Aggregate using mean to find average and store in a variable = new_df.
new_df = data[['fuel_type_code_pudl','fuel_cost_per_unit_burned']].groupby('fuel_type_code_pudl').mean()

#sort values in ascending order so lowest value comes first in the dataframe
new_df.sort_values('fuel_cost_per_unit_burned')

fuel_cost_per_unit_burned
fuel_type_code_pudl                           
gas                                  13.659397
other                                18.253856
waste                                19.518122
coal                                 67.421830
oil                                 168.877086
nuclear                            4955.157002

In [12]:
#Get first value
answer = new_df.sort_values('fuel_cost_per_unit_burned', ascending=True).iloc[0]
answer

fuel_cost_per_unit_burned    13.659397
Name: gas, dtype: float64

## **Question 6: Standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?**

In [16]:
data[['fuel_mmbtu_per_unit']].describe()

fuel_mmbtu_per_unit
count         29523.000000
mean              8.492111
std              10.600220
min               0.000001
25%               1.024000
50%               5.762694
75%              17.006000
max             341.260000

Read from the dataframe above:
- std = 10.60
- 75th percentile = 17.01

## **Question 7: Skewness and kurtosis for the fuel quantity burned in two decimal places**

In [26]:
skewness = data['fuel_qty_burned'].skew()
skewness.round(2)

15.85

In [27]:
kurtosis = data['fuel_qty_burned'].kurtosis()
kurtosis.round(2)

651.37

## **Question 4: For the fuel type 'coal', percentage change in the fuel_cost_per_unit_burned in 1998 compared to 1994**

In [19]:
#create a new dataframe for fuel type 'coal' and store in variable = coal.
coal = data[data['fuel_type_code_pudl'] == 'coal']

#extract two dataframes from coal with report years 1994 and 1998
coal_1994 = coal[coal['report_year'] == 1994]
coal_1998 = coal[coal['report_year'] == 1998]

#find the sum of 'fuel_cost_per_unit_burned' for the two dataframes.
#sum_1994 = sum of coal in 1994, sum_1998 = sum of coal in 1998
sum_1994 = coal_1994['fuel_cost_per_unit_burned'].sum(axis=0)
sum_1998 = coal_1998['fuel_cost_per_unit_burned'].sum(axis=0)

#percentage change = (new - old)/old * 100

percent_change = (sum_1998 - sum_1994)/sum_1994 * 100
percent_change.round()

-21.0

## **Question 8: Feature with missing values, total number of missing value and percentage of missing rows as a factor of the total rows in 3dp.**

In [20]:
#feature with missing values
data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

`fuel_unit` is the only feature with missing values.

In [22]:
#total number of missing values
total_miss = data.isnull().sum().sum()
total_miss

180

In [25]:
#percentage of missing rows to total number of rows
total_rows = data.shape[0]

percentage = total_miss / total_rows * 100
percentage.round(3)

0.61

**Feature: fuel_unit, Total: 180, Percent: 0.61**

## **Question 6: Second and Third Lowest correlation with the Fuel Cost Per Unit Burned.**

In [29]:
#find correlation using 'pearson' method: standard correlation coefficient
correlation = data.corr(method='pearson')
correlation

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

In [30]:
#sort values in the required column to ascending
correlation.sort_values('fuel_cost_per_unit_burned')

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
report_year                           0.093323     1.000000         0.012952   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
report_year                             -0.110853                   0.013599   
fuel_cost_per_unit_burned               -0.010034                   1.000000   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_mmbtu                              -0.000109   
fuel_cost_per_unit_delivered                      1.000000   
report_year                                      -0.014043   
fuel_cost_per_unit_burned                         0.011007   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_mmbtu                      1.000000  
fuel_cost_per_unit_delivered            -0.000109  
report_year                              0.010261  
fuel_cost_per_unit_burned               -0.000437

From the dataframe above, the 2nd and 3rd lowest correlation values to `fuel_cost_per_unit_burned` is **fuel_qty_burned** and **fuel_mmbtu_per_unit.**